File: 496-next-greater-element-i/496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> st = new Stack<>();
        Map<Integer, Integer> mp = new HashMap<>();
        int[] ans = new int[nums1.length];
        Arrays.fill(ans, -1);
        for(int i = 0; i < nums1.length; i++)
            mp.put(nums1[i], i);
        
        for(int i = 0; i < nums2.length; i++){
            if(st.isEmpty())
                st.add(i);
            else if(nums2[i] <= nums2[st.peek()])
                st.add(i);
            else{
                while(!st.isEmpty() && nums2[i] > nums2[st.peek()]){
                    if(mp.containsKey(nums2[st.peek()])){
                        ans[mp.get(nums2[st.peek()])] = nums2[i];
                    }
                    st.pop();
                }
                st.add(i);
            }
        }
        return ans;
    }
}

File: 496-next-greater-element-i/496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        st = []
        d = {nums1[i] : i for i in range(len(nums1))}
        # print(d)
        res = [-1]*len(nums2)
        ans = [-1]*len(nums1)
        for i in range(len(nums2)):
            if len(st) == 0:
                st.append(i)
            elif nums2[i] <= nums2[st[-1]]:
                st.append(i)
            else:
                while len(st) != 0 and nums2[i] > nums2[st[-1]]:
                    res[st[-1]] = nums2[i]
                    if(nums2[st[-1]] in d):
                        ans[d[nums2[st[-1]]]] = nums2[i]
                    st.pop()
                st.append(i)
        # print(nums2)
        # print(res)
        # print(nums1)
        return ans